File: sorting/sort.py
import time
import random


a = [3, 44, 38, 5, 47, 151, 36, 26, 27, 2, 46, 61, 19, 50, 50, 48]


def execution_time(f):
    start_time = time.time()
    f()
    print('Execution time is {} seconds.'.format(time.time() - start_time))


def bubble_sort():
    for i in range(len(a) - 1):
        for j in range(i + 1, len(a)):
            if a[i] > a[j]:
                a[i], a[j] = a[j], a[i]


def insertion_sort():
    for i in range(1, len(a)):
        j = i
        while j > 0 and a[j - 1] > a[i]:
            j -= 1
        tmp = a[i]
        for k in range(i, j, -1):
            a[k] = a[k - 1]
        a[j] = tmp


def _merge_sort(a, left, right):
    if left < right:
        mid = (left + right) // 2
        _merge_sort(a, left, mid)
        _merge_sort(a, mid + 1, right)
        tmp = [0] * len(a)
        curr, i, j = left, left, mid + 1
        while i <= mid or j <= right:
            if i > mid:
                tmp[curr] = a[j]
                curr += 1
                j += 1
            elif j > right:
                tmp[curr] = a[i]
                curr += 1
                i += 1
            elif a[i] < a[j]:
                tmp[curr] = a[i]
                curr += 1
                i += 1
            else:
                tmp[curr] = a[j]
                curr += 1
                j += 1
        # print(tmp)    # uncomment this line to see a pretty visualization
        for i in range(left, right + 1):
            a[i] = tmp[i]


def merge_sort():
    _merge_sort(a, 0, len(a) - 1)


def heapify(root, end):
    left = root * 2 + 1
    right = root * 2 + 2
    max = a[root]
    pos = root
    if left < end and max < a[left]:
        max = a[left]
        pos = left
    if right < end and max < a[right]:
        max = a[right]
        pos = right
    if pos != root:
        a[root], a[pos] = a[pos], a[root]
        heapify(pos, end)


def heap_sort():
    for i in range(len(a) - 1, -1, -1):
        heapify(i, len(a))
    for i in range(len(a) - 1, 0, -1):
        a[0], a[i] = a[i], a[0]
        heapify(0, i - 1)


def _quick_sort(a, left, right):
    i, j = left, right
    pivot = a[random.randint(left, right)]
    while i <= j:
        while a[i] < pivot:
            i += 1
        while a[j] > pivot:
            j -= 1
        if i <= j:
            a[i], a[j] = a[j], a[i]
            i += 1
            j -= 1
    if j > left:
        _quick_sort(a, left, j)
    if i < right:
        _quick_sort(a, i, right)


def quick_sort():
    _quick_sort(a, 0, len(a) - 1)


def radix_sort():
    BASE = 13
    exp = 1
    while BASE ** exp <= max(a) * BASE:
        tmp = [[] for _ in range(BASE)]
        for i in a:
            tmp[i % (BASE ** exp) // (BASE ** (exp - 1))].append(i)
        curr = 0
        for i in tmp:
            for j in i:
                a[curr] = j
                curr += 1
        exp += 1


print(a)
execution_time(quick_sort)
print(a)

File: sorting/sort.cpp
#include <cmath>
#include <iostream>
#include <vector>

int a[] = {3, 44, 38, 5, 47, 151, 36, 26, 27, 2, 46, 61, 19, 50, 50, 48};
int n = sizeof(a) / sizeof(*a);

void bubbleSort() {
    for (int i = 0; i < n - 1; ++i) {
        for (int j = i + 1; j < n; ++j) {
            if (a[i] > a[j]) {
                std::swap(a[i], a[j]);
            }
        }
    }
}

void selectionSort() {
    for (int i = 0; i < n - 1; ++i) {
        int min = INT_MAX, pos = -1;
        for (int j = i + 1; j < n; ++j) {
            if (a[j] < min) {
                min = a[j];
                pos = j;
            }
        }
        if (a[i] > a[pos]) {
            std::swap(a[i], a[pos]);
        }
    }
}

void insertionSort() {
    for (int i = 1; i < n; ++i) {
        int j = i;
        while (j > 0 && a[j - 1] > a[i]) {
            --j;
        }
        int tmp = a[i];
        for (int k = i; k > j; --k) {
            a[k] = a[k - 1];
        }
        a[j] = tmp;
    }
}

void mergeSort(int left, int right) {
    if (left < right) {
        int mid = (left + right) / 2;
        mergeSort(left, mid);
        mergeSort(mid + 1, right);
        std::vector<int> tmp(n, 0);
        int curr = left;
        int i = left;
        int j = mid + 1;
        while (i <= mid || j <= right) {
            if (i > mid) {
                tmp[curr++] = a[j++];
            } else if (j > right) {
                tmp[curr++] = a[i++];
            } else if (a[i] > a[j]) {
                tmp[curr++] = a[j++];
            } else {
                tmp[curr++] = a[i++];
            }
        }
        for (int k = left; k <= right; ++k) {
            a[k] = tmp[k];
        }
    }
}

void heapify(int root, int end) {
    int ptr = root, left = 2 * root + 1, right = 2 * root + 2;
    if (left < end && a[left] > a[ptr]) {
        ptr = left;
    }
    if (right < end && a[right] > a[ptr]) {
        ptr = right;
    }
    if (ptr != root) {
        std::swap(a[root], a[ptr]);
        heapify(ptr, end);
    }
}

void heapSort() {
    for (int i = n - 1; i >= 0; --i) {
        heapify(i, n);
    }
    for (int i = n - 1; i >= 0; --i) {
        std::swap(a[0], a[i]);
        heapify(0, i);
    }
}

void quickSort(int left, int right) {
    if (left >= right) return;
    int pivot = a[left + rand() % (right - left + 1)];
    int i = left, j = right;
    while (i <= j) {
        while (a[i] < pivot) ++i;
        while (a[j] > pivot) --j;
        if (i <= j) {
            std::swap(a[i++], a[j--]);
        }
    }
    quickSort(left, j);
    quickSort(i, right);
}

void radixSort() {
    int BASE = 10, exp = 1, max = INT_MIN;
    for (int i : a)
        if (max < i) max = i;
    while (pow(BASE, exp) <= max * BASE) {
        std::vector<std::vector<int>> tmp(BASE, std::vector<int>());
        for (int i : a)
            tmp[i % (int)pow(BASE, exp) / (int)pow(BASE, exp - 1)].push_back(i);
        int curr = 0;
        for (std::vector<int> i : tmp)
            for (int j : i)
                a[curr++] = j;
        ++exp;
    }
}

void printArray() {
    if (n <= 0) {
        return;
    }
    std::cout << "[";
    for (int i = 0; i < n - 1; ++i) {
        std::cout << a[i] << ", ";
    }
    std::cout << a[n - 1] << "]" << std::endl;
}

int main() {
    printArray();
    selectionSort();
    printArray();

    return 0;
}
